upcloud/gateway.go

```diff
@@ -57,3 +105,82 @@ type GatewayRouter struct {
+	Name          string           `json:"name,omitempty"`
+	StaticNetwork string           `json:"static_network,omitempty"`
+	Type          GatewayRouteType `json:"type,omitempty"`
+}
+
+type GatewayTunnel struct {
+	Name             string                        `json:"name,omitempty"`
+	LocalAddress     GatewayTunnelLocalAddress     `json:"local_address,omitempty"`
+	RemoteAddress    GatewayTunnelRemoteAddress    `json:"remote_address,omitempty"`
+	IPSec            GatewayTunnelIPSec            `json:"ipsec,omitempty"`
+	OperationalState GatewayTunnelOperationalState `json:"operational_state,omitempty"`
+	CreatedAt        time.Time                     `json:"created_at,omitempty"`
+	UpdatedAt        time.Time                     `json:"updated_at,omitempty"`
+}
+
+type GatewayTunnelLocalAddress struct {
+	// Name of the UpCloud gateway address; should correspond to the name of one of the gateway address structs
+	Name string `json:"name,omitempty"`
+}
+
+type GatewayTunnelRemoteAddress struct {
+	// Address is a remote peer address VPN will connect to; must be global non-private unicast IP address.
+	Address string `json:"address,omitempty"`
+}
+
+type GatewayTunnelIPSec struct {
+	// Tunnel IPSec authentication object
+	Authentication GatewayTunnelIPSecAuth `json:"authentication,omitempty"`
+	// IKE SA rekey time in seconds
+	RekeyTime int `json:"rekey_time,omitempty"`
+	// IKE child SA rekey time in seconds
+	ChildRekeyTime int `json:"child_rekey_time,omitempty"`
+	// Delay before sending Dead Peer Detection packets if no traffic is detected, in seconds
+	DPDDelay int `json:"dpd_delay,omitempty"`
+	// Timeout period for DPD reply before considering the peer to be dead, in seconds
+	DPDTimeout int `json:"dpd_timeout,omitempty"`
+	// Maximum IKE SA lifetime in seconds
+	IKELifetime int `json:"ike_lifetime,omitempty"`
+	// List of Phase 1: Proposal algorithms
+	Phase1Algortihms []GatewayIPSecAlgorithm `json:"phase1_algorithms,omitempty"`
+	// List of Phase 1 integrity algorithms
+	Phase1IntegrityAlgorithms []GatewayIPSecIntegrityAlgorithm `json:"phase1_integrity_algorithms,omitempty"`
+	// List of Phase 1 Diffie-Hellman group numbers
+	Phase1DHGroupNumbers []int `json:"phase1_dh_group_numbers,omitempty"`
+	// List of Phase 2: Security Association algorithms
+	Phase2Algortihms []GatewayIPSecAlgorithm `json:"phase2_algorithms,omitempty"`
+	// List of Phase 2 integrity algorithms
+	Phase2IntegrityAlgorithms []GatewayIPSecIntegrityAlgorithm `json:"phase2_integrity_algorithms,omitempty"`
+	// List of Phase 2 Diffie-Hellman group numbers
+	Phase2DHGroupNumbers []int `json:"phase2_dh_group_numbers,omitempty"`
+}
+
+type GatewayTunnelIPSecAuth struct {
+	Authentication GatewayIPSecAuthType `json:"authentication,omitempty"`
+	// PSK is a user-provided pre-shared key.
+	// Note that this field is only meant to be used when providing API with your pre-shared key; it will always be empty in API responses
+	PSK string `json:"psk,omitempty"`
```

Comment:
Alternatively I could remove `PSK` field from here and create a separate type for create/modify request that has both `Authentication` and `PSK` fields.
I went with this option because IMO we already have a lot of types around gateways and the less types we have, the easier it is to use SDK. But I don't have a strong opinion on this.

Reply:
I think we already use both approaches (different type in `request` package vs. shared type from `upcloud` package) elsewhere as well and this seems quite clean to me 👍